File: app/src/main/java/me/ykrank/s1next/data/api/ApiCacheProvider.java
package me.ykrank.s1next.data.api;

import java.util.concurrent.TimeUnit;

import io.reactivex.Single;
import io.rx_cache2.DynamicKey;
import io.rx_cache2.DynamicKeyGroup;
import io.rx_cache2.EvictDynamicKey;
import io.rx_cache2.EvictDynamicKeyGroup;
import io.rx_cache2.LifeCache;
import io.rx_cache2.ProviderKey;
import io.rx_cache2.Reply;

/**
 * RxCache provide cache for retrofit
 * Created by ykrank on 2017/4/22.
 */
public interface ApiCacheProvider {

    @ProviderKey("forum_groups_wrapper")
    @LifeCache(duration = 1, timeUnit = TimeUnit.MINUTES)
    Single<String> getForumGroupsWrapper(Single<String> oWrapper, DynamicKey user, EvictDynamicKey evictDynamicKey);

    @ProviderKey("threads_wrapper")
    @LifeCache(duration = 1, timeUnit = TimeUnit.MINUTES)
    Single<String> getThreadsWrapper(Single<String> oWrapper, DynamicKeyGroup user, EvictDynamicKeyGroup evictDynamicKey);


    @ProviderKey("posts_wrapper")
    @LifeCache(duration = 30, timeUnit = TimeUnit.MINUTES)
    Single<Reply<String>> getPostsWrapper(Single<String> oWrapper, DynamicKeyGroup page, EvictDynamicKeyGroup evictDynamicKey);
}

File: app/src/main/java/me/ykrank/s1next/data/api/ApiCacheProvider.kt
package me.ykrank.s1next.data.api

import com.github.ykrank.androidtools.data.CacheParam
import com.github.ykrank.androidtools.data.Resource
import kotlinx.coroutines.flow.Flow
import me.ykrank.s1next.data.api.model.Rate
import me.ykrank.s1next.data.api.model.wrapper.ForumGroupsWrapper
import me.ykrank.s1next.data.api.model.wrapper.PostsWrapper
import me.ykrank.s1next.data.api.model.wrapper.ThreadsWrapper

/**
 * RxCache provide cache for retrofit
 * Created by ykrank on 2017/4/22.
 */
interface ApiCacheProvider {
    /**
     * 板块列表
     */
    suspend fun getForumGroupsWrapper(
        param: CacheParam? = null
    ): Flow<Resource<ForumGroupsWrapper>>

    /**
     * 板块详情/主题列表
     */
    suspend fun getThreadsWrapper(
        forumId: String?,
        typeId: String?,
        page: Int,
        param: CacheParam? = null
    ): Flow<Resource<ThreadsWrapper>>

    /**
     * 主题详情/帖子列表
     */
    suspend fun getPostsWrapper(
        threadId: String?,
        page: Int,
        authorId: String? = null,
        ignoreCache: Boolean = false,
        onRateUpdate: ((pid: Int, rate: List<Rate>) -> Unit)? = null,
    ): Flow<Resource<PostsWrapper>>

    /**
     * 评分
     */
    suspend fun getPostRates(
        threadId: String?,
        postId: Int,
    ): Resource<List<Rate>>

}
